### PR TITLE
mgr/dashboard: Added graphs to visualise perf counter time sequences

### PR DESCRIPTION
**Pull request by Saumay-Agrawal**
- Tables can now be sorted on the basis of space usage, r/w activities, daemon status etc.
- Tables can now be searched for specific details

Signed-off-by: Saumay Agrawal <saumay.agrawal@gmail.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
@Saumay-Agrawal  needs rebase and a screenshot would be helpful.

File: src/pybind/mgr/dashboard/osd_perf.html [hunk 1]
```diff
@@ -121,7 +121,7 @@
 
 <section class="content">
 
-    <table>
+    <!-- <table>
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf.html]: why you are commenting out this table?

**Comment by tchaikov** [comment]
rw-perf

@Saumay-Agrawal could you explain a little bit how this diagram helps us understand the performance of the OSD ?

**Comment by jcsp** [comment]
@tchaikov the osd read/write latency bits are not new in this PR -- they were a rough thing that I added a while ago.

**Comment by Saumay-Agrawal** [comment]
The changes added in this commit are as follows:

**Addition of get methods to module.py**

- Added a method get_counter_allosd(self, path) which consolidates the time sequences for the perf counter specified as the argument 'path', given by get_counter() method for all the osds in a single JSON object containing {osd_id: time_sequence} pairs.
- Added a method get_counter_stats(self, path) which uses the data collected by get_counter_allosd() and process it to return the statistical data (minimum, maximum, average and stdevs) about the perf counter specified as the argument 'path'.
- Added perf_graphs_prototypes(self, path) which shows the various sample graphs of the given perf counter. It exposes the webpage perf_graphs_prototypes.html at localhost:41000/perf_graph_prototypes/{perf_counter}. These graphs serves the purpose of a general machinery, and will make implementation of performance graphs easier for future developers.
- Added get_perf_graph_data(self) and osd_perf_graph(self) to expose the webpage containing read and write latency perf counters on the dashboard.

general_prototypes_snapshot


**General UI features of the graphs**

- The perf counter values for all OSDs are represented by thin white lines.
- The minimum and maximum perf counter values are respresented by thick green colored lines.
- The average perf counter values are represented by thick orange colored lines.
- The standard deviations on both sides of the average value are represented by thick black lines.
- On hovering the mouse pointer on points plotted on the graph, exact value for that dataset is shown in a tooltip along with its label. For eg. OSD id can be identified on seeing a spike on white line by hovering over it.
- If two or more datasets intersect at the same point, i.e. they have same value for a time instant, then the tool tip lists all the values along with their labels.

latency_perf_graph_snapshots


**OSD Performance Graph (Webpage visualizing Latencies)**
- This page is linked to the Cluster/OSDs page on the dashboard.
- This page contains four graphs. Two graphs show the minimum, maximum and average values for the read and write latencies each. And the other two graphs show the distribution of read and write latencies in terms of process and prepare latencies.

File: src/pybind/mgr/dashboard/osd_perf.html (hunk 1 above)
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf.html]: @Saumay-Agrawal why you are commenting out these two heapmap and adding two new ones?

File: src/pybind/mgr/dashboard/osd_perf.html [hunk 2]
```diff
@@ -121,20 +121,6 @@
 
 <section class="content">
 
-    <!-- <table>
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf.html]: please squash this change into 65bd5188c00a8dd68823fb5db2d85c5a573e75aa

File: src/pybind/mgr/dashboard/module.py [hunk 3]
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
...
+
+                      for osd_id in range(0,osd_count):
+                            values.append(data[osd_id][i][1])
+
+                      result.append(
+                            [epoch, values])
 
                 return result
 
+            @cherrypy.expose
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: why would you want to expose this method?

File: src/pybind/mgr/dashboard/module.py [hunk 4]
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
...
+
+                return result
+
+            @cherrypy.expose
+            def perf_graph(self, path):
+                template = env.get_template("perf_graphs.html")
+                toplevel_data = self._toplevel_data()
+
+                content_data = {
+                    "avg" : self.get_counter_avg(path),
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: no need to `get_counter_data()` repeatedly, please store it locally. and get the avg and stddev in a single run.

File: src/pybind/mgr/dashboard/module.py [hunk 5]
```diff
@@ -751,20 +752,110 @@ def get_perf_schema(self, **args):
...
+                result = []
                 for osd in osd_map['osds']:
                     osd_id = "{0}".format(osd['osd'])
                     counter_data = global_instance().get_counter('osd', osd_id, path)
-                    result[osd['osd']] = counter_data[path]
+                    result.append(counter_data[path])
+
+                return result
+
+            #Method to find the average of consolidated data
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: s/Method to /

File: src/pybind/mgr/dashboard/perf_graphs.html [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+                return data;
+            };
+	
+    var min_values = convert_timeseries(content_data["min"]);
+    var max_values = convert_timeseries(content_data["max"]);
+    var avg_values = convert_timeseries(content_data["avg"]);
+    var stdev_values = convert_timeseries(content_data["stdev"]);
+
+	var minmax = document.getElementById("minmax-chart");
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/perf_graphs.html]: could you be more consistent when using space and tab for indent?

File: src/pybind/mgr/dashboard/perf_graphs.html [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+				borderColor: "rgb(62, 149, 205)",
+				fillColor: "rgba(62, 149, 205, 0.2)",
+			}, { 
+				data: max_values,
+				label: "Max latency",
+				borderColor: "rgb(142, 94, 162)",
+				fillColor: "rgba(142, 94, 162, 0.2)",
+			}
+			]
+
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/perf_graphs.html]: remove empty lines.

File: src/pybind/mgr/dashboard/static/AdminLTE-2.3.7/plugins/datatables/datatables_custom_darkUI.css [hunk 8]
```diff
@@ -1,357 +0,0 @@
-/*
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/static/AdminLTE-2.3.7/plugins/datatables/datatables_custom_darkUI.css]: squash this change into 65bd5188c00a8dd68823fb5db2d85c5a573e75aa. in other words, please avoid fix-the-fix changes in a single PR.

File: src/pybind/mgr/dashboard/module.py [hunk 9]
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
...
+                    epoch_set[epoch].append(osd_id)
+                max_sync = 0
+                for epoch in epoch_set:
+                    sync_count = len(epoch_set[epoch])
+                    if sync_count > max_sync:
+                        ref_osd = epoch_set[epoch][0]
+                        max_sync = sync_count
+
+                for i in range(0,len(data[ref_osd])):
+
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: remove the empty line.

File: src/pybind/mgr/dashboard/module.py [hunk 10]
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
...
+                for tseq in result:
+                    
+                    length = len(tseq[1])
+                    avg = sum(tseq[1])/length
+                    counter_stats_data["avg"].append([tseq[0], avg])
+                    diff = [(x-avg)**2 for x in tseq[1]]
+                    var = sum(diff)/length
+                    stdev = math.sqrt(var)
+                    counter_stats_data["stdev"].append([tseq[0], stdev])
+                    counter_stats_data["stdevs"]["plus_onehalf"].append([tseq[0], avg+(1.5*stdev)])
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: what are these plots for: plus_onehalf, plus_half, ...?

File: src/pybind/mgr/dashboard/module.py [hunk 11]
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
...
+                content_data = self.get_counter_stats(path)
+
+                return template.render(
+                    ceph_version=global_instance().version,
+                    toplevel_data=json.dumps(toplevel_data, indent=2),
+                    content_data=json.dumps(content_data, indent=2),
+                )
+
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: again, why shall we expose this method?

File: src/pybind/mgr/dashboard/module.py [hunk 12]
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
...
+                    osd_id = "osd-" + osd_id
+                    result[osd_id] = counter_data[path]
+                return result
+
+            # Method to get the overall statistical distribution of a given perf counter.
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def get_counter_stats(self, path):
+                
+                data = self.get_counter_allosd(path)
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: could you use a better name than data? for example `counters`?

File: src/pybind/mgr/dashboard/module.py [hunk 13]
```diff
@@ -751,6 +752,133 @@ def get_perf_schema(self, **args):
...
+                    "stdevs" : {
+                        "plus_onehalf" : [],
+                        "plus_half" : [],
+                        "minus_half" : [],
+                        "minus_onehalf" : []
+                    },
+                    "osds_out_sync" : []
+                    }
+
+                result = []
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/module.py]: `time_seqs`.

File: src/pybind/mgr/dashboard/module.py (hunk 5 above)
Comment (Saumay-Agrawal) [reply src/pybind/mgr/dashboard/module.py]: It is fixed in the latest commit. Perhaps, I should squash all the changes related to graphs in a single commit.

File: src/pybind/mgr/dashboard/module.py (hunk 3 above)
Comment (Saumay-Agrawal) [reply src/pybind/mgr/dashboard/module.py]: This function is no longer there in the final commit. Again, I would like your opinion on squashing all the changes related to graphs and TS in a single commit.

File: src/pybind/mgr/dashboard/module.py (hunk 11 above)
Comment (Saumay-Agrawal) [reply src/pybind/mgr/dashboard/module.py]: Again, This function is no longer there in the final commit. I would like your opinion on squashing all the changes related to graphs and TS in a single commit.

File: src/pybind/mgr/dashboard/module.py (hunk 4 above)
Comment (Saumay-Agrawal) [reply src/pybind/mgr/dashboard/module.py]: I admit you are right. I realised the same while I was working on the graphs, so I did replace the entire set of separate methods into a single method get_counter_stats() which stores the data of get_counter_allosd() one time, and then processes it to return the needed values of average, stdevs etc.

File: src/pybind/mgr/dashboard/module.py (hunk 10 above)
Comment (Saumay-Agrawal) [reply src/pybind/mgr/dashboard/module.py]: These plots are to refer to the standard deviations around average, plus_onehalf = Average + (1.5 *stdev), plus_half = Average + (0.5*stdev), minus_half  = Average - (0.5*stdev) and minus_onehalf = average - (1.5*stdev). This way distribution of latency values on both sides of the average value can be seen for the performance of OSDs.

File: src/pybind/mgr/dashboard/module.py (hunk 5 above)
Comment (tchaikov) [reply src/pybind/mgr/dashboard/module.py]: @Saumay-Agrawal commits should be self-contained and be atomic. please do so.

File: src/pybind/mgr/dashboard/osd_perf.html [hunk 14]
```diff
@@ -121,26 +121,34 @@
...
+        <div class="col-md-4">
+            <h3>Reads</h3>
+            <table id="heatmap_read">
+            </table>
+        </div>
+            
+    </div>
+
+<hr>
+
```
Comment (Saumay-Agrawal) [inline src/pybind/mgr/dashboard/osd_perf.html]: @tchaikov  I tried to remove the empty lines, but they are not showing up here. Ironically, they are removed in my local repo and the file is committed too. Can you please help me out in this?

**Comment by tchaikov** [comment]
@Saumay-Agrawal could you drop unrelated commits (d6d502c043d230ded31656564ee42493fa12add5 and a2a6a4d) from this PR?

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+{% block content %}
+
+<section class="content-header">
+    <h1>
+        OSD Performance Graphs
+    </h1>
+</section>
+
+<section class="content">
+
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: remove the empty lines.

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 16]
```diff
@@ -0,0 +1,412 @@
+{% extends "base.html" %}
+
+{% block content %}
+
+<section class="content-header">
+    <h1>
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: nit, put this tag in a single line.

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+        OSD Performance Graphs
+    </h1>
+</section>
+
+<section class="content">
+	
+	<h3>Read Latencies</h3>
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: please just use space for indent.

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+    <div class="row">
+		<div id="read1" class="col-md-6">
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: ditto.

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+
+	var refresh = function() {
+	    $.get("/get_perf_graph_data/", function(data) {
+	        content_data = data;
+	        plot_graphs();
+	        setTimeout(refresh, 5000);
+	    });
+	}; 
+
+	var convert_timeseries = function(source_series)
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: indent.

File: src/pybind/mgr/dashboard/osd_perf_graph.html [hunk 20]
```diff
@@ -0,0 +1,412 @@
...
+
+    var plot_graphs = function() {
+    	plot_read();
+        plot_read_dist();
+        plot_write();
+        plot_write_dist();
+    };
+
+    var plot_read = function() {
+
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/osd_perf_graph.html]: remove empty line.

File: src/pybind/mgr/dashboard/servers.html [hunk 21]
```diff
@@ -79,6 +82,15 @@
...
     </div>
+
+    <script>
+      $(function () {
+        $("#servers-table").DataTable({
+            paging: false
+        });
+      });
+    </script>
+
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/servers.html]: remove empty line.

File: src/pybind/mgr/dashboard/base.html [hunk 22]
```diff
@@ -31,6 +33,7 @@
     <script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.17.1/moment.min.js"></script>
     <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.4.0/Chart.min.js"></script>
 
+    <script src="/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.min.js"></script>
```
Comment (nrdmn) [inline src/pybind/mgr/dashboard/base.html]: This script has already been added in de9e6a7ef0476f62ea4e1116d057e2f79089c185 but in a different line

**Comment by nrdmn** [comment]
This PR conflicts with #17119. I've prepared a patch to resolve this.

**Comment by Saumay-Agrawal** [comment]
@nrdmn Thanks for the patch. I've merged it and resolved the conflicts with #17119.

**Comment by tchaikov** [comment]
@Saumay-Agrawal please don't use "merge pull request" to include @nrdmn 's pull request, you should have use "rebased and merge". normally we don't want PRs to include "merge" commit. you might want to rebase your change locally and repush.

File: src/pybind/mgr/dashboard/base.html [hunk 23]
```diff
@@ -17,9 +17,9 @@
     <link rel="stylesheet"
           href="{{ url_prefix }}/static/AdminLTE-2.3.7/dist/css/AdminLTE.min.css">
     <link rel="stylesheet"
-          href="{{ url_prefix }}/static/AdminLTE-2.3.7/dist/css/skins/skin-blue.min.css">
+          href="/static/AdminLTE-2.3.7/dist/css/skins/skin-blue.min.css">
     <link rel="stylesheet"
-          href="{{ url_prefix }}/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.css">
+          href="/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.css">
```
Comment (jcsp) [inline src/pybind/mgr/dashboard/base.html]: this has removed the {{url_prefix}}, it should remain (and also appear in the datatables url below)

**Comment by jcsp** [comment]
I get this backtrace when visiting /osd_perf_graph/:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/usr/lib/python2.7/dist-packages/cherrypy/lib/encoding.py", line 217, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/cherrypy/_cpdispatch.py", line 61, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 953, in osd_perf_graph
    content_data = self.get_perf_graph_data()
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 943, in get_perf_graph_data
    data = self.get_counter_stats(counter)
  File "/home/jspray/ceph.bravo/src/pybind/mgr/dashboard/module.py", line 879, in get_counter_stats
    epoch = str(counters[osd_id][0][0])
IndexError: list index out of range
```

**Comment by jcsp** [comment]
This seems to be abandoned -- probably not sensible to merge any of it given that it presumably doesn't cope with large numbers of OSDs

**Pull request closed without merging**